src/main/java/com/openlattice/chronicle/services/delete/DataDeletionManagerImpl.java

```diff
@@ -0,0 +1,225 @@
+            DataApi chronicleDataApi = apiClient.getDataApi();
+
+            // get required entity set ids
+            UUID studiesESID = commonTasksManager.getEntitySetId( organizationId, CHRONICLE, STUDIES, STUDY_ES );
+            UUID participantsESID = commonTasksManager.getParticipantEntitySetId( organizationId, studyId );
+            checkNotNullUUIDs( Sets.newHashSet( studiesESID, participantsESID ) );
+
+            // these entity set ids will be null if the respective app modules have not been installed for the organization
+            UUID appDataESID = commonTasksManager
+                    .getEntitySetId( organizationId, CHRONICLE_DATA_COLLECTION, APPDATA, DATA_ES );
+            UUID preprocessedDataESID = commonTasksManager
+                    .getEntitySetId( organizationId, CHRONICLE_DATA_COLLECTION, PREPROCESSED_DATA,
+                            PREPROCESSED_DATA_ES );
+            UUID devicesESID = commonTasksManager
+                    .getEntitySetId( organizationId, CHRONICLE_DATA_COLLECTION, DEVICE, DEVICES_ES );
+            UUID answersESID = commonTasksManager
+                    .getEntitySetId( organizationId, CHRONICLE_SURVEYS, ANSWER, ANSWERS_ES );
+
+            // ensure study exists
+            UUID studyEntityKeyId = Preconditions
+                    .checkNotNull( commonTasksManager.getStudyEntityKeyId( organizationId, studyId ),
+                            "study must exist" );
+
+            // get a set of all participants to remove:
+            Set<UUID> participantsToRemove = new HashSet<>();
+            if ( participantId.isPresent() ) {
+                // if participantId: add to set
+                UUID participantEntityKeyId = commonTasksManager
+                        .getParticipantEntityKeyId( organizationId, studyId, participantId.get() );
+                if ( participantEntityKeyId == null ) {
+                    throw new Exception(
+                            "unable to delete participant " + participantId + ": participant does not exist." );
+                }
+                participantsToRemove.add( participantEntityKeyId );
+            } else {
+                // if no participant Id: load all participants and add to set
+                userDataApi
+                        .loadEntitySetData( participantsESID, FileType.json, userToken )
+                        .forEach( entity -> entity.get( ID_FQN ).forEach( personId ->
+                                        participantsToRemove.add( UUID.fromString( personId.toString() ) )
+                                )
+                        );
+            }
+
+            // Be super careful here that the mapping is one-to-one:
+            // don't delete neighbors that might have other neighbors/participants
+
+            Set<UUID> srcNeighborSetIds = new HashSet<>(), dstNeighborSetIds = Sets.newHashSet();
+
+            if ( devicesESID != null )
```

Comment:
Good point. Actually I think it should be sufficient to do ImmutableSet.of(....) without the Sets.filter since ImmutableSet.of (...) will reject nulls

ref: https://guava.dev/releases/19.0/api/docs/src-html/com/google/common/collect/ImmutableSet.html#line.55

Reply:
I think `ImmutableSet.of( null )` will throw an exception

Reply:
Oh yeah actually in that case the code I wrote would throw an NPE.

But actually now that I think about it, these entity set ids should never be null, right? Are there any conditions where they *would* be null?

Reply:
You are right:
<img width="796" alt="Screen Shot 2020-11-05 at 10 47 24 AM" src="https://user-images.githubusercontent.com/15723547/98283313-593fcf00-1f54-11eb-9cc4-42af7e2cb4c0.png">

Reply:
Yes, these entity set ids could be null. In the app configs world, we have split up chronicle into 3 components: a core component, data collection and surveys. If an organization has not installed all three components, then some of these entity set ids will be null.